src/procedures/proc_degree.c

```diff
@@ -0,0 +1,445 @@
+	//--------------------------------------------------------------------------
+
+	if(Map_Get(config, SI_ConstStringVal("relation"), &tmp)) {
+		if(SI_TYPE(tmp) != T_STRING) {
+			return false;
+		}
+
+		*rel_type = tmp.stringval;
+		matches++;
+	}
+
+	//--------------------------------------------------------------------------
+	// get destination label
+	//--------------------------------------------------------------------------
+
+	if(Map_Get(config, SI_ConstStringVal("destination"), &tmp)) {
+		if(SI_TYPE(tmp) != T_STRING) {
+			return false;
+		}
+
+		*dest_label = tmp.stringval;
+		matches++;
+	}
+
+	// make sure config map doesn't contains unexpected keys
+	if(Map_KeyCount(config) > matches) {
+		// map contains unmatched keys
+		// e.g.
+		// {'sources': 'Person', 'unexpected_key': 4}
+		return false;
+	}
+
+	return true;
+}
+
+// procedure invoke callback
+ProcedureResult Proc_DegreeInvoke
+(
+	ProcedureCtx *ctx,    // procedure context
+	const SIValue *args,  // invocation args
+	const char **yield    // output(s)
+) {
+	// expecting a single map argument
+	if(array_len((SIValue *)args) != 1) return PROCEDURE_ERR;
+
+	// configuration
+	GRAPH_EDGE_DIR dir;      // edge direction
+	const char *src_label;   // src label
+	const char *rel_type;    // edge relationship type
+	const char *dest_label;  // destination label
+
+	// parse configuration
+	if(!parse_arguments(args[0], &src_label, &dir, &rel_type, &dest_label)) {
+		// failed to parse input configuration
+		// emit an error and return
+		ErrorCtx_RaiseRuntimeException(EMSG_PROC_INVALID_ARGUMENTS, "algo.degree");
+		return PROCEDURE_ERR;
+	}
+
+	Schema *s;
+	GraphContext *gc        = QueryCtx_GetGraphCtx();
+	Graph        *g         = GraphContext_GetGraph(gc);
+	bool          transpose = (dir == GRAPH_EDGE_DIR_INCOMING);
+
+	Delta_Matrix S = NULL;  // source label matrix
+	Delta_Matrix R = NULL;  // relation adjacency matrix
+	Delta_Matrix D = NULL;  // destination label matrix
+
+	//--------------------------------------------------------------------------
+	// get source label matrix
+	//--------------------------------------------------------------------------
+
+	if(src_label != NULL) {
+		s = GraphContext_GetSchema(gc, src_label, SCHEMA_NODE);
+		if(s != NULL) {
+			S = Graph_GetLabelMatrix(g, Schema_GetID(s));
+		} else {
+			S = Graph_GetZeroMatrix(g);
+		}
+	}
+
+	//--------------------------------------------------------------------------
+	// get relationship type matrix
+	//--------------------------------------------------------------------------
+
+	if(rel_type != NULL) {
+		s = GraphContext_GetSchema(gc, rel_type, SCHEMA_EDGE);
+		if(s != NULL) {
+			R = Graph_GetRelationMatrix(g, Schema_GetID(s), transpose);
+			// error in case relationship contains tensors
+			// TODO: support tensors
+			if(Graph_RelationshipContainsMultiEdge(g, Schema_GetID(s))) {
+				ErrorCtx_RaiseRuntimeException("'algo.degree' can't run against a graph containing tensors");
```

Comment:
add messge to error_msgs.h
are you sure it is safe to raise exception?

Reply:
You're right, I've changed that to `ErrorCtx_SetError`